### PR TITLE
[LibOS] Support `FIONREAD` ioctl on sockets

### PR DESCRIPTION
**Pull request by boryspoplawski**
## Description of the changes 
Fixes https://github.com/gramineproject/gramine/issues/673

---
This change is [Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)

### CODE REVIEW

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: 0 of 3 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/684#-N5N7-Vd5gKODAAaKaNP:-N5N7-Ve5-72IxfZaOjM:b-x9vfen) (no related file):*
TODO: add a test

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/684#-N5N75R1-huh27r2sRG6:-N5N75R1-huh27r2sRG7:bszd73v) (no related file):*
based on https://github.com/gramineproject/gramine/pull/681 so blocking until it's merged

___

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @boryspoplawski)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/684#-N5N75R1-huh27r2sRG6:-N5Z5sO13nNfUliEJfES:b-69is6c) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

based on https://github.com/gramineproject/gramine/pull/681 so blocking until it's merged
</blockquote></details>

#681 was merged.

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/684#-N5Z6ysQ92o4RLhRJf2J:-N5Z6ysQ92o4RLhRJf2K:b-xak97i) (no related file):*
Jenkins failed.

___
___
*[`libos/src/fs/socket/fs.c` line 76 at r1](https://reviewable.io/reviews/gramineproject/gramine/684#-N5Z6ESYDKiVnBwzz7Ro:-N5Z6ESYDKiVnBwzz7Rp:b-sk34fv) ([raw file](https://github.com/gramineproject/gramine/blob/0a7112ab9a875c1618638465a6c2b50a5743d5bb/libos/src/fs/socket/fs.c#L76)):*
> ```C
>     stat->st_blksize = PAGE_SIZE;
> 
>     /* XXX: maybe set `st_size` - query PAL for pending size? Otoh nothing seems to be using it. */
> ```

Why did you remove this comment?

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: 2 of 3 files reviewed, 3 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @dimakuv)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/684#-N5Z6ysQ92o4RLhRJf2J:-N5_3WrVAcmkrjVmZjc1:b-896fix) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Jenkins failed.
</blockquote></details>

Done.

___
___
*[`libos/src/fs/socket/fs.c` line 76 at r1](https://reviewable.io/reviews/gramineproject/gramine/684#-N5Z6ESYDKiVnBwzz7Ro:-N5_3_rk7z4f5RFqiZxC:b-fvqtkz) ([raw file](https://github.com/gramineproject/gramine/blob/0a7112ab9a875c1618638465a6c2b50a5743d5bb/libos/src/fs/socket/fs.c#L76)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Why did you remove this comment?
</blockquote></details>

Because it's invalid - this function should be used only by `*stat` functions, which do not return socket pending size.

**Review by dimakuv — APPROVED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: all files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required)

**Review by mkow — DISMISSED** [review]
> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: all files reviewed, 1 unresolved discussion (waiting on @boryspoplawski)

**Review by dimakuv — APPROVED** [review]
> Reviewed 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: all files reviewed, 1 unresolved discussion (waiting on @boryspoplawski)

**Review by mkow — APPROVED** [review]
> Reviewed 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: all files reviewed, 1 unresolved discussion (waiting on @boryspoplawski)

**Review by dimakuv — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: all files reviewed, 2 unresolved discussions, not enough approvals from different teams (1 more required, approved so far: ITL) (waiting on @boryspoplawski)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/684#-N66u7noAK_W6UAAlZqv:-N66u7npDWQsfWsaguZ0:b-8fv9cm) (no related file):*
Jenkins failed like this:
```
19:38:23  ../libos/src/fs/socket/fs.c: In function ‘get_socket_pending_size’:
19:38:23  ../libos/src/fs/socket/fs.c:136:15: error: implicit declaration of function ‘DkStreamAttributesQueryByHandle’; did you mean ‘PalStreamAttributesQueryByHandle’? [-Werror=implicit-function-declaration]
19:38:23       int ret = DkStreamAttributesQueryByHandle(pal_handle, &attr);
19:38:23                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
19:38:23                 PalStreamAttributesQueryByHandle
19:38:23  cc1: all warnings being treated as errors
```

___

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: 4 of 7 files reviewed, 1 unresolved discussion, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @mkow)

*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/684#-N5N7-Vd5gKODAAaKaNP:-N69nBAADI_mJ63lCLzl:b-896fix) (no related file):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

TODO: add a test
</blockquote></details>

Done.

___
*[a discussion](https://reviewable.io/reviews/gramineproject/gramine/684#-N66u7noAK_W6UAAlZqv:-N69nC9mDPRqeLJKqFw2:bounkh4) (no related file):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Jenkins failed like this:
```
19:38:23  ../libos/src/fs/socket/fs.c: In function ‘get_socket_pending_size’:
19:38:23  ../libos/src/fs/socket/fs.c:136:15: error: implicit declaration of function ‘DkStreamAttributesQueryByHandle’; did you mean ‘PalStreamAttributesQueryByHandle’? [-Werror=implicit-function-declaration]
19:38:23       int ret = DkStreamAttributesQueryByHandle(pal_handle, &attr);
19:38:23                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
19:38:23                 PalStreamAttributesQueryByHandle
19:38:23  cc1: all warnings being treated as errors
```
</blockquote></details>

How did that build locally for me... Must have forgot to fetch or somerhing

___

**Review by mkow — CHANGES_REQUESTED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @boryspoplawski and @dimakuv)

___
*[`libos/test/regression/tcp_msg_peek.c` line 82 at r4](https://reviewable.io/reviews/gramineproject/gramine/684#-N69sfa6EtwJ9K0kICUc:-N69sfa79u0F8RV4lJgR:b-hic2po) ([raw file](https://github.com/gramineproject/gramine/blob/1ee331147853f5180b9cc2c8e72fcc45ba33ed1b/libos/test/regression/tcp_msg_peek.c#L82)):*
> ```C
>     ssize_t x = CHECK(read(pipefd, &c, sizeof(c)));
>     if (x != 1) {
>         CHECK(-1);
> ```

I'd prefer calling `errx()` directly in such places (`CHECK(-1)` is a bit weird).
___
*[`libos/test/regression/tcp_msg_peek.c` line 143 at r4](https://reviewable.io/reviews/gramineproject/gramine/684#-N69tHYRBeGUc6jc8Cqj:-N69tHYRBeGUc6jc8Cqk:b-kcc36t) ([raw file](https://github.com/gramineproject/gramine/blob/1ee331147853f5180b9cc2c8e72fcc45ba33ed1b/libos/test/regression/tcp_msg_peek.c#L143)):*
> ```C
>     count = client_recv(s, buf, sizeof(buf), /*flags=*/0);
>     if (count != sizeof(g_buffer)) {
>         errx(1, "recv with MSG_PEEK returned less that than available: %zd", count);
> ```

Something went wrong here?



_Code quote:_
```C
less that than available
```

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: all files reviewed, 3 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @mkow)

___
*[`libos/test/regression/tcp_msg_peek.c` line 82 at r4](https://reviewable.io/reviews/gramineproject/gramine/684#-N69sfa6EtwJ9K0kICUc:-N6A2NEmCIg-fqX8KfoW:bs82693) ([raw file](https://github.com/gramineproject/gramine/blob/1ee331147853f5180b9cc2c8e72fcc45ba33ed1b/libos/test/regression/tcp_msg_peek.c#L82)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

I'd prefer calling `errx()` directly in such places (`CHECK(-1)` is a bit weird).
</blockquote></details>

Basically, if I have some value to print then I write a custom `err`, but if I just want to get the line where the error happened I use this. Can change if you insist, but it won't give any extra value (won't be more useful).
___
*[`libos/test/regression/tcp_msg_peek.c` line 143 at r4](https://reviewable.io/reviews/gramineproject/gramine/684#-N69tHYRBeGUc6jc8Cqj:-N6A3mjt7d5ISA3MsHkm:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/1ee331147853f5180b9cc2c8e72fcc45ba33ed1b/libos/test/regression/tcp_msg_peek.c#L143)):*
<details><summary><i>Previously, mkow (Michał Kowalczyk) wrote…</i></summary><blockquote>

Something went wrong here?
</blockquote></details>

Done.

**Review by mkow — APPROVED** [review]
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: all files reviewed, 1 unresolved discussion, not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv)

___
*[`libos/test/regression/tcp_msg_peek.c` line 82 at r4](https://reviewable.io/reviews/gramineproject/gramine/684#-N69sfa6EtwJ9K0kICUc:-N6AFT-W-ezGURxcZm35:b3j6vkl) ([raw file](https://github.com/gramineproject/gramine/blob/1ee331147853f5180b9cc2c8e72fcc45ba33ed1b/libos/test/regression/tcp_msg_peek.c#L82)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

Basically, if I have some value to print then I write a custom `err`, but if I just want to get the line where the error happened I use this. Can change if you insist, but it won't give any extra value (won't be more useful).
</blockquote></details>

Ok, the argument with the line number is convincing.

**Review by dimakuv — CHANGES_REQUESTED** [review]
> Reviewed 2 of 3 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: all files reviewed, 2 unresolved discussions, not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @boryspoplawski)

___
*[`libos/test/regression/tcp_msg_peek.c` line 92 at r5](https://reviewable.io/reviews/gramineproject/gramine/684#-N6CSPACBwOcEu_uNAxj:-N6CSPACBwOcEu_uNAxk:b6pbvoo) ([raw file](https://github.com/gramineproject/gramine/blob/0accc53999361113e936df85ab3545209ce33384/libos/test/regression/tcp_msg_peek.c#L92)):*
> ```C
>         .sin_port = htons(PORT),
>         .sin_addr = {
>             /* TODO: remove this once Ubuntu 18.04 is deprecated. */
> ```

Could you remind again what's this problem?
___
*[`libos/test/regression/tcp_msg_peek.c` line 147 at r5](https://reviewable.io/reviews/gramineproject/gramine/684#-N6CSTe_4hj7-Hkjgcah:-N6CSTea5W4S_nr4imik:b-42fhpy) ([raw file](https://github.com/gramineproject/gramine/blob/0accc53999361113e936df85ab3545209ce33384/libos/test/regression/tcp_msg_peek.c#L147)):*
> ```C
>     count = client_recv(s, buf, sizeof(buf), /*flags=*/0);
>     if (count != sizeof(g_buffer)) {
>         errx(1, "recv with MSG_PEEK returned less than available: %zd", count);
> ```

`with` -> `without`

**Review by boryspoplawski — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: 7 of 8 files reviewed, 2 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), "fixup! " found in commit messages' one-liners (waiting on @dimakuv and @mkow)

___
*[`libos/test/regression/tcp_msg_peek.c` line 92 at r5](https://reviewable.io/reviews/gramineproject/gramine/684#-N6CSPACBwOcEu_uNAxj:-N6DRRXw1Aeh5-C0O9Kp:b-knaa4y) ([raw file](https://github.com/gramineproject/gramine/blob/0accc53999361113e936df85ab3545209ce33384/libos/test/regression/tcp_msg_peek.c#L92)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

Could you remind again what's this problem?
</blockquote></details>

No idea. Ubuntu 18.04 (well gcc there) complains this field is not initialized, but it is (implicitly to 0)
___
*[`libos/test/regression/tcp_msg_peek.c` line 147 at r5](https://reviewable.io/reviews/gramineproject/gramine/684#-N6CSTe_4hj7-Hkjgcah:-N6DSErl0U7P--OMgPKB:b-896fix) ([raw file](https://github.com/gramineproject/gramine/blob/0accc53999361113e936df85ab3545209ce33384/libos/test/regression/tcp_msg_peek.c#L147)):*
<details><summary><i>Previously, dimakuv (Dmitrii Kuvaiskii) wrote…</i></summary><blockquote>

`with` -> `without`
</blockquote></details>

Done.

**Review by dimakuv — DISMISSED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: all files reviewed, all discussions resolved, "fixup! " found in commit messages' one-liners

___
*[`libos/test/regression/tcp_msg_peek.c` line 92 at r5](https://reviewable.io/reviews/gramineproject/gramine/684#-N6CSPACBwOcEu_uNAxj:-N6DXXOe1xFi1KtlmOsV:b-uhl6l4) ([raw file](https://github.com/gramineproject/gramine/blob/0accc53999361113e936df85ab3545209ce33384/libos/test/regression/tcp_msg_peek.c#L92)):*
<details><summary><i>Previously, boryspoplawski (Borys Popławski) wrote…</i></summary><blockquote>

No idea. Ubuntu 18.04 (well gcc there) complains this field is not initialized, but it is (implicitly to 0)
</blockquote></details>

Ok, right, this is the "something something" bug.

**Review by mkow — DISMISSED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: all files reviewed, all discussions resolved, "fixup! " found in commit messages' one-liners

**Review by dimakuv — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by mkow — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gramine/684)* status: :shipit: complete! all files reviewed, all discussions resolved
